
## `this.` 를 안 붙이면 `ReferenceError`가 뜨는데 위에 선언을 했음에도 불구하고 왜 에러가 뜨는지 원인 찾아보기


#### Original Code

```js
var person = {
    name: "홍길동",
    birthday: "030219",
    age: 30,
    pId: "1234567",
    fullId: function() {
        return this.birthday + this.pId;
    }
};

person.fullId() // 0302191234567
person.fullId;  // function () { return this.birthday + this.pId; }

//생성자(constructor)를 이용한 객체 생성
var day = new Date(); // new 연산자를 사용하여 Date 타입의 객체를 생성함.
console.log("올해는 " + day.getFullYear() + "년입니다.");
```

#### Edited Code

```js
var person = {
    name: "홍길동",
    birthday: "030219",
    age: 30,
    pId: "1234567",
    fullId: function() {
        return birthday + pId;
    }
};
```

위 코드에 따라 `person` 객체 안에 있는 `fullID` 함수를 아래처럼 호출하면 `ReferenceError`가 발생된다.

```js
person.fullId();
```

RESULT

```bash
C:\Users\Eunice\OneDrive - Sogang\바탕 화면\git\ai_webdev\js_ts\this_reference_error\8_object.js:7
        return birthday + pId;
        ^

ReferenceError: birthday is not defined
    at Object.fullId (C:\Users\Eunice\OneDrive - Sogang\바탕 화면\git\ai_webdev\js_ts\this_reference_error\8_object.js:7:9)
    at Object.<anonymous> (C:\Users\Eunice\OneDrive - Sogang\바탕 화면\git\ai_webdev\js_ts\this_reference_error\8_object.js:10:8)
    at Module._compile (node:internal/modules/cjs/loader:1254:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1308:10)
    at Module.load (node:internal/modules/cjs/loader:1117:32)
    at Module._load (node:internal/modules/cjs/loader:958:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47

Node.js v18.15.0
```

단, 아래와 같이 함수를 호출하면 정상 결과가 나타난다.

```js
person.fullId;
```

RESULT

```
올해는 2023년입니다.
```

---


#### 설명

ReferenceError 발생 원인
- 객체는 현재 범위에서 존재하지 않거나 (non-existing), 
- 초기화되지 않은 (not initialized) 변수를 참조했을 때 발생하는 에러

this
- 객체 내부에서 사용될 때 현재 해당 객체(current object)를 가리킨다

코드 안에서 this는 person 객체 안에서 사용되었으니
`this`는 `person`을 참조하게 된다.
-->  this.birthday는 person의 birthday 속성을을 가리킨다.


###### 괄호 '()'를 사용하여 `fullId` 함수 호출

```js
person.fullId()
```

- 위 방식대로 `fullID` 함수를 호출하면 이 함수가 person 객체의 함수로 호출된다. 
- 원래 코드에서 `return this.birthday + this.pId` 를 쓰면 `this`는 `person` 객체를 가리키고, `this.birthday` 와 `this.pId`는 `person`의 `birthday`와 `pID` 프로포티를 가리킨다.
- 이 경우 `return birthday + pId`와 같이 `this.` 를 안 붙이고 접근을 시도하면 JS 인터프리터는 `birthday`와 `pId` 를 `person` 객체의 property가 아닌 `fullId`  함수 안에 정의된 변수들이라고 가정할 것이다. `fullId` 안에서 `birthday`와 `pId` 라는 변수가 선언되지 않았으니 정의가 없으며`undefined` 가 된다. 이 때문에 `ReferenceError`가 발생된다.
	- 자바스크립트에서 `this.`를 붙여서 범위(scope) 또는 문맥(context)를 알려주지 않으면 인터프리터는 현재 범위(current scope)부터 상위 범위(parent scope), 그리고 전역 범위(global scope)까지 탐색하여 해당 변수를 찾는다. (`fullId` 함수는 `person` 객체 범위 안에 있으니 `fullId`의 현재 범위는 함수 안의 범위이고 상위 범위는 전역 범위(global)이다. `birthday` 와 `pId`는 함수 안에 그리고 전역적으로 정의되어 있지 않기 때문에 인터프리터가 변수들을 찾지 못하는 상황이 되며 ReferenceError의 원인이 된다.)

따라서 `person.fullId()`로 호출할 때 `this.birthday`와  `this.pId`를 사용해야 `person` 객체의 property를 참조하면 `ReferenceError` 가 발생하지 않는다.

###### 괄호 '()' 없이 `fullId` 함수 호출

```js
person.fullId;
```

- 위 방식의 경우 `person.fullId`는 함수를 호출 또는 실행시키지 않고 참조만 하는 방식이다.
- 함수를 호출하지 않고 참조만 하기 때문에 함수의 문맥(context) 즉 this의 값이 중요하지 않다. 따라서 `fullId` 함수 안에서 `birthday` 와 `pId` 속성에 접근하려는 시도를 하지 않으며, `return birthday + pId`를 사용해도 ReferenceError가 발생하지 않는다.

---


#### 요약

`person.fullId();` 
- person 객체의 메서드로서 함수를 호출한다. 
- `fullId` 함수 내부의 `this` 키워드는 `person` 객체를 가리키며, `this.birthday`와 `this.pId`에 접근할 수 있다.

`person.fullId;`
- 함수 자체를 반환하며 호출하지 않다
- 함수는 참조만 되고 실행되지 않기 때문에 오류가 발생하지 않다.

